cohort-engine/src/main/java/com/ibm/cohort/engine/cqfruler/MeasureEvaluation.java

```diff
@@ -81,7 +81,29 @@ public MeasureReport evaluatePatientMeasure(Measure measure, Context context, St
                 MeasureReport.MeasureReportType.INDIVIDUAL, isSingle, includeEvaluatedResources);
     }
 
-    @SuppressWarnings("unchecked")
+    public MeasureReport evaluatePatientListMeasure(Measure measure, Context context, List<String> patientIds, boolean includeEvaluatedResources) {
+	    logger.info("Generating patient-list report");
+
+	    List<Patient> patients = toPatients(patientIds);
+	    boolean isSingle = false;
+	    return evaluate(measure, context, patients, MeasureReport.MeasureReportType.SUBJECTLIST, isSingle, includeEvaluatedResources);
+    }
+
+	private List<Patient> toPatients(List<String> patientIds) {
+		List<Patient> patients = new ArrayList<>();
+
+		for (String patientId : patientIds) {
+			Iterable<Object> patientRetrieve = provider.retrieve(PATIENT, "id", patientId, PATIENT, null, null, null, null, null, null, null, null);
```

Comment:
This is going to cause a separate round trip to the data system for every patient ID in the list which seems excessive to me. I realize our options are somewhat limited with the existing interface definition. In the original CQF Ruler code, the MeasureEvaluation client would pass in a DaoRegistry (HAPI specific) class that would allow the evaluator to build a search query. The CQL Evaluator doesn't have the flexibility. There, you either get all patients in the target system or all patients for a specific practitioner, both of which are single requests to the data provider for data.

I spent some time wondering whether or not we could cheat the system and pass "_id" as the code path and the list of IDs as dummy Code objects, but I'm pretty sure that won't work. Whenever the backend sees "_id" it tries to convert the request to a read vs. a search. Nothing in the existing DataProvider interface/impls will support a general search for a list of patients. If we want this to perform better, we would need to switch to our own data provider interface that is more flexible than what the CQL backend version provides.

Reply:
Definitely a valid point.
I went ahead and created a ticket in the backlog for this: [WFFHCOHORT-474](https://jira.wh-sdlc.watson-health.ibm.com/browse/WFFHCOHORT-474)